From 9ca1552737fb758daeff055e563a027040c3dd74 Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Fri, 5 Apr 2024 22:03:38 +0200
Subject: [PATCH] panfrost: Add the BO containing fragment program descriptor
 to the batch

On pre-Valhall HW, the fragment shader metadata was part of the RSD
(renderer state descriptor), which was emitted at draw time, but
Valhall introduces a shader program descriptor containing only the
shader information, and this one is emitted at shader preparation
time.

If we don't add the FS state BO to batch, we might end up with a batch
being executed after the shader object has been destroyed, leading to
page faults when the GPU tries to access the shader program descriptor.

We make the panfrost_batch_add_bo() unconditional since it gracefully
handles the NULL case (which will happen on v7-).

Fixes: 087b63cb0771 ("panfrost: Allow uploading fragment SPDs")
Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Erik Faye-Lund <erik.faye-lund@collabora.com>
Reviewed-by: Antonino Maniscalco <antonino.maniscalco@collabora.com>
---
 src/gallium/drivers/panfrost/pan_cmdstream.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/gallium/drivers/panfrost/pan_cmdstream.c b/src/gallium/drivers/panfrost/pan_cmdstream.c
index 034d77d2bb74f..82383949e94d4 100644
--- a/src/gallium/drivers/panfrost/pan_cmdstream.c
+++ b/src/gallium/drivers/panfrost/pan_cmdstream.c
@@ -646,6 +646,7 @@ panfrost_emit_frag_shader_meta(struct panfrost_batch *batch)
    struct panfrost_compiled_shader *ss = ctx->prog[PIPE_SHADER_FRAGMENT];
 
    panfrost_batch_add_bo(batch, ss->bin.bo, PIPE_SHADER_FRAGMENT);
+   panfrost_batch_add_bo(batch, ss->state.bo, PIPE_SHADER_FRAGMENT);
 
    struct panfrost_ptr xfer;
 
-- 
GitLab

From ce7b37ca613dbb141484fa40702a607a6a82c823 Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Wed, 24 Apr 2024 15:28:43 +0200
Subject: [PATCH] pan/kmod: Fix a syncobj leak in the panthor backend

Make sure we release the syncobj attached to the BO if the object is
sharable.

Fixes: 97f6a62f7ef8 ("pan/kmod: Add a backend for panthor")
Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
Reviewed-by: Erik Faye-Lund <erik.faye-lund@collabora.com>
Acked-by: Constantine Shablia <constantine.shablya@collabora.com>
Reviewed-by: Antonino Maniscalco <antonino.maniscalco@collabora.com>
---
 src/panfrost/lib/kmod/panthor_kmod.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/panfrost/lib/kmod/panthor_kmod.c b/src/panfrost/lib/kmod/panthor_kmod.c
index 0c7fbc61f2465..4404b6b1cad86 100644
--- a/src/panfrost/lib/kmod/panthor_kmod.c
+++ b/src/panfrost/lib/kmod/panthor_kmod.c
@@ -309,6 +309,12 @@ err_free_bo:
 static void
 panthor_kmod_bo_free(struct pan_kmod_bo *bo)
 {
+   struct panthor_kmod_bo *panthor_bo =
+      container_of(bo, struct panthor_kmod_bo, base);
+
+   if (!bo->exclusive_vm)
+      drmSyncobjDestroy(bo->dev->fd, panthor_bo->sync.handle);
+
    drmCloseBufferHandle(bo->dev->fd, bo->handle);
    pan_kmod_dev_free(bo->dev, bo);
 }
-- 
GitLab

From 2097c30705568f3ea2500a629a0be4e7f1a5c727 Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@collabora.com>
Date: Tue, 30 Apr 2024 13:56:41 +0200
Subject: [PATCH] pan/kmod: Make default allocator thread-safe

Allocations targeting a pan_kmod_dev can happen concurrently, so we
need the pan_kmod_dev allocator to be thread-safe.

ralloc() is not thread-safe, and we don't really need a hierarchical
allocator in this context anyway, so let's just switch to calloc/free
instead.

Fixes: d95ec56f8c68 ("panfrost: Abstract kernel driver operations")
Reported-by: Eric Smith <eric.smith@collabora.com>
Signed-off-by: Boris Brezillon <boris.brezillon@collabora.com>
---
 src/panfrost/lib/kmod/pan_kmod.c | 37 ++++++++------------------------
 1 file changed, 9 insertions(+), 28 deletions(-)

diff --git a/src/panfrost/lib/kmod/pan_kmod.c b/src/panfrost/lib/kmod/pan_kmod.c
index df376a136c432..e8a656e7436a4 100644
--- a/src/panfrost/lib/kmod/pan_kmod.c
+++ b/src/panfrost/lib/kmod/pan_kmod.c
@@ -7,6 +7,7 @@
 #include <string.h>
 #include <xf86drm.h>
 
+#include "util/u_memory.h"
 #include "util/macros.h"
 #include "pan_kmod.h"
 
@@ -31,28 +32,19 @@ static void *
 default_zalloc(const struct pan_kmod_allocator *allocator, size_t size,
                UNUSED bool transient)
 {
-   return rzalloc_size(allocator, size);
+   return os_calloc(1, size);
 }
 
 static void
 default_free(const struct pan_kmod_allocator *allocator, void *data)
 {
-   return ralloc_free(data);
+   os_free(data);
 }
 
-static const struct pan_kmod_allocator *
-create_default_allocator(void)
-{
-   struct pan_kmod_allocator *allocator =
-      rzalloc(NULL, struct pan_kmod_allocator);
-
-   if (allocator) {
-      allocator->zalloc = default_zalloc;
-      allocator->free = default_free;
-   }
-
-   return allocator;
-}
+static const struct pan_kmod_allocator default_allocator = {
+   .zalloc = default_zalloc,
+   .free = default_free,
+};
 
 struct pan_kmod_dev *
 pan_kmod_dev_create(int fd, uint32_t flags,
@@ -64,11 +56,8 @@ pan_kmod_dev_create(int fd, uint32_t flags,
    if (!version)
       return NULL;
 
-   if (!allocator) {
-      allocator = create_default_allocator();
-      if (!allocator)
-         goto out_free_version;
-   }
+   if (!allocator)
+      allocator = &default_allocator;
 
    for (unsigned i = 0; i < ARRAY_SIZE(drivers); i++) {
       if (!strcmp(drivers[i].name, version->name)) {
@@ -82,9 +71,6 @@ pan_kmod_dev_create(int fd, uint32_t flags,
       }
    }
 
-   if (allocator->zalloc == default_zalloc)
-      ralloc_free((void *)allocator);
-
 out_free_version:
    drmFreeVersion(version);
    return dev;
@@ -93,12 +79,7 @@ out_free_version:
 void
 pan_kmod_dev_destroy(struct pan_kmod_dev *dev)
 {
-   const struct pan_kmod_allocator *allocator = dev->allocator;
-
    dev->ops->dev_destroy(dev);
-
-   if (allocator->zalloc == default_zalloc)
-      ralloc_free((void *)allocator);
 }
 
 struct pan_kmod_bo *
-- 
GitLab


